datadriven/src/sgpp/datadriven/algorithm/GridFactory.cpp

```diff
@@ -52,68 +51,275 @@ sgpp::base::Grid *GridFactory::createGrid(const sgpp::base::GeneralGridConfigura
+                           multiplicatorsPerLevel.at(i), interactions);
     }
+    addOneDimensionalInteractions(geometryConf.dim.at(i), offsetsPerLevel.at(i), interactions);
   }
 
-  for (size_t i = 0; i < geodimX; i++) {
-    for (size_t j = 0; j < geodimY-1; j++) {
-      std::vector<size_t> ydir = std::vector<size_t>();
+  // Add empty interaction
+  interactions.insert(std::set<size_t>());
+}
 
-      ydir.push_back(i+j*geodimX);
+void sgpp::datadriven::GridFactory::addOneDimensionalInteractions(
+    std::vector<int64_t>& imageDimensions, size_t offset, std::set<std::set<size_t>>& vec) const {
+  size_t factor = 1;
+  for (auto dimension : imageDimensions) {
+    factor *= dimension;
+  }
 
-      ydir.push_back(i+(j+1)*geodimX);
+  for (size_t i = offset; i < factor + offset; i++) {
+    vec.insert({i});
+  }
+}
 
-      vec.push_back(ydir);
+void sgpp::datadriven::GridFactory::addChildParentInteraction(
+    std::vector<double>& ratio, size_t numberOfAxes, size_t currentAxis,
+    std::vector<int64_t>& childPosition, std::vector<int64_t>& parentPosition,
+    std::vector<size_t>& parentMultiplicators, std::vector<size_t>& childMultiplicators,
+    size_t parentOffset, size_t childOffset, std::set<std::set<size_t>>& interactions) const {
+  // Check if all axes has been set
+  if (currentAxis < numberOfAxes) {
+    // Calculate first and last parent on this axis
+    size_t begin = static_cast<size_t>(ratio.at(currentAxis) *
+                                       static_cast<double>(childPosition.at(currentAxis)));
+    double end = ratio.at(currentAxis) * static_cast<double>(childPosition.at(currentAxis) + 1);
+
+    // Iterate over all parents' positions
+    for (size_t i = begin; static_cast<double>(i) < end; i++) {
+      parentPosition.at(currentAxis) = i;
+      addChildParentInteraction(ratio, numberOfAxes, currentAxis + 1, childPosition, parentPosition,
+                                parentMultiplicators, childMultiplicators, parentOffset,
+                                childOffset, interactions);
     }
+  } else {
+    auto interaction = std::set<size_t>();
+    interaction.insert(getDataIndex(numberOfAxes, childMultiplicators, childPosition) +
+                       childOffset);
+    interaction.insert(getDataIndex(numberOfAxes, parentMultiplicators, parentPosition) +
+                       parentOffset);
+    interactions.insert(interaction);
   }
+}
 
-  // 1d vector for all dimensions
-  for (size_t i = 0; i < geodimX*geodimY; i++) {
-    std::vector<size_t> tmp = std::vector<size_t>();
-    tmp.push_back(i);
-    vec.push_back(tmp);
+size_t sgpp::datadriven::GridFactory::getDataIndex(size_t numberOfDimensions,
+                                                   std::vector<size_t>& multiplicators,
+                                                   std::vector<int64_t>& position) const {
+  size_t index = 0;
+  for (size_t i = 0; i < numberOfDimensions; i++) {
+    index += multiplicators.at(i) * position.at(i);
```

Comment:
Bounds checking with `at` necessary? Otherwise use `[]`.

Reply:
The other `at`s are necessary? `[]` is twice as fast as `at`, so I use the latter only for debugging purposes.